src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -32527,5 +32611,203 @@ class B2 : A<int>
             var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8, references: new[] { ref0 });
             comp.VerifyDiagnostics();
         }
+
+        [WorkItem(27967, "https://github.com/dotnet/roslyn/issues/27967")]
+        [Fact]
+        public void UnannotatedTypeArgument_Interface()
+        {
+            var source0 =
+@"public interface I<T>
+{
+}
+public class B : I<object[]>
+{
+}
+public class C : I<C>
+{
+}";
+            var comp0 = CreateCompilation(source0, parseOptions: TestOptions.Regular7);
+            comp0.VerifyDiagnostics();
+            var ref0 = comp0.EmitToImageReference();
+
+            var source =
+@"class Program
+{
+    static void F(B x)
+    {
+        I<object[]?> a = x;
+        I<object[]> b = x;
+    }
+    static void F(C y)
+    {
+        I<C?> a = y;
+        I<C> b = y;
+    }
+}";
+            var comp = CreateCompilation(source, references: new[] { ref0 }, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics();
+        }
+
+        [WorkItem(27967, "https://github.com/dotnet/roslyn/issues/27967")]
+        [Fact]
+        public void UnannotatedTypeArgument_BaseType()
+        {
+            var source0 =
+@"public class A<T>
+{
+}
+public class B : A<object[]>
+{
+}
+public class C : A<C>
+{
+}";
+            var comp0 = CreateCompilation(source0, parseOptions: TestOptions.Regular7);
+            comp0.VerifyDiagnostics();
+            var ref0 = comp0.EmitToImageReference();
+
+            var source =
+@"class Program
+{
+    static void F(B x)
+    {
+        A<object[]?> a = x;
+        A<object[]> b = x;
+    }
+    static void F(C y)
+    {
+        A<C?> a = y;
+        A<C> b = y;
+    }
+}";
+            var comp = CreateCompilation(source, references: new[] { ref0 }, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void UnannotatedTypeArgument_Interface_Lookup()
+        {
+            var source0 =
+@"public interface I<T>
+{
+    void F(T t);
+}
+public interface I1 : I<object>
+{
+}";
+            var comp0 = CreateCompilation(source0, parseOptions: TestOptions.Regular7);
+            comp0.VerifyDiagnostics();
+            var ref0 = comp0.EmitToImageReference();
+
+            var source =
+@"interface I2 : I<object>
+{
+}
+class Program
+{
+    static void F(I1 i1, I2 i2, object x, object? y)
+    {
+        i1.F(x);
+        i1.F(y);
+        i2.F(x);
+        i2.F(y); // warn
```

Comment:
This one is interesting. Perhaps more of a nullable design question, but is there going to be a way for a user to specify that `I2.F` takes a `T?`? Seems like a user would want to be able to override this behavior.

Reply:
An implicit implementation of `I.F` could be defined as `F(T?)` since the interface method is null-oblivious, but `interface I2` could not redefine `I.F`.

---
In reply to: [198655729](https://github.com/dotnet/roslyn/pull/28028#discussion_r198655729) [](ancestors = 198655729)